### PR TITLE
Harden Sentry diagnostics scrubbing (FE + BE)

### LINKED ISSUE #3787: Sentry diagnostics scrubbing: close leak vectors and fix isolated-client wiring before merging qa/sentry-settings
## Summary

The `qa/sentry-settings` branch hardens Sentry diagnostics scrubbing and newly wires two browser integrations (`httpContextIntegration`, `browserApiErrorsIntegration`) plus expanded backend scrubbing. A `max`-effort multi-agent code review of `git diff main...HEAD` (8 files) surfaced **10 verified issues** that should be closed before this branch merges.

Two themes:

1. **Scrubbing lags what the new integrations now attach.** The new integrations populate request/transaction fields (Referer header, full URL, query strings) that the `beforeSend`/`beforeSendTransaction` scrubbers don't yet cover, so route identifiers and auth tokens can reach Sentry unredacted — the exact class of leak this branch is meant to prevent.
2. **The isolated Sentry client is never bound as the global current client.** Both new integrations resolve their client via the global `getClient()`, which is `undefined` in this setup, so async-callback errors are silently dropped and no transaction events are ever produced (making the transaction scrubber dead code).

> Full technical detail, exact field paths, verifier evidence, and reproduction scenarios are captured in the private engineering memory referenced at the bottom of this issue.

## Scope

Changed files under review:
- `src/plugins/core/enableDiagnostics.ts` (frontend)
- `lib/onetime/initializers/setup_diagnostics.rb` (backend)
- `spec/unit/onetime/initializers/setup_diagnostics_spec.rb`
- + supporting scrubbers / docs

### A. Data-leak gaps (scrubbing does not cover new fields)

- [ ] **A1 — Frontend URL query params not scrubbed.** `httpContextIntegration` sets `event.request.url = location.href`; the URL scrubbers redact route/path patterns but not sensitive query-param *names* (e.g. `key`, `token`, `secret`, `passphrase`). Auth tokens on token-bearing routes reach Sentry. **Fix:** add query-string param-name scrubbing to the frontend URL scrubber (parity with backend `scrub_url`).
- [ ] **A2 — Backend `Referer` header not scrubbed.** `scrub_event_urls` scrubs `request.url` / `event.transaction` / `contexts.request.url` but not `event.request.headers['Referer']`, which sentry-ruby captures verbatim. A Referer carrying a route identifier ships unredacted. **Fix:** scrub `request.headers['Referer']` (and equivalent) through the same path scrubber.
- [ ] **A3 — Frontend `Referer` header not scrubbed.** `httpContextIntegration` attaches `document.referrer` as `event.request.headers.Referer` (not gated on `sendDefaultPii`); neither `beforeSend` nor `beforeSendTransaction` scrubs `event.request.headers`. **Fix:** scrub `event.request.headers` in both handlers.
- [ ] **A4 — Transaction spans omit `http.query`.** The frontend transaction scrubber loops `['url','http.url','url.full']` but not `http.query`, which the fetch/xhr instrumentation populates with the raw query string. Backend already scrubs `data['http.query']`. **Fix:** add `http.query` to the frontend span-data scrub loop.

### B. Isolated-client wiring (integrations partly non-functional)

- [ ] **B1 — Async-callback errors silently dropped.** `browserApiErrorsIntegration` wraps timer / event-listener / XHR callbacks, but its error path reports via the global `captureException()` → `getClient()`, which is `undefined` here (only `scope.setClient(client)` is called). Errors in `setTimeout`/`addEventListener`/`requestAnimationFrame`/XHR callbacks — which Vue's `errorHandler` also never sees — produce **no Sentry event**. **Fix:** bind the client as current (`setCurrentClient`) or route capture through the isolated client explicitly.
- [ ] **B2 — Zero transactions recorded; transaction scrubber is dead code.** Same root cause: `browserTracingIntegration` resolves its client via global `getClient()`, so tracing yields only non-recording spans and `sendSpanEnvelope` short-circuits. Despite `tracesSampleRate` being set, no performance transactions are sent and `beforeSendTransaction` never runs. **Fix:** same as B1; then re-verify the transaction scrubber actually executes.

> ⚠️ Fixing B1/B2 activates the transaction pipeline — so A4 and finding D2 below become live leaks the moment the client is bound. Land the scrubbing fixes (A/D) together with the wiring fix.

### C. Regex parity is falsely claimed

- [ ] **C1 — Backend/frontend identifier patterns diverge despite a "mirrors" comment.** Backend `IDENTIFIER_TEXT_PATTERN` is `\b(?:{62}|{31})\b` (word-boundary, case-sensitive); frontend `VERIFIABLE_ID_PATTERN` is `{62}` case-insensitive, unanchored. Consequences: (a) backend leaves a 62-char id intact when it abuts a word char or when a 63+ char run appears; (b) frontend misses legacy 31-char identifiers (still supported, `MIN_IDENTIFIER_LENGTH=20`). **Fix:** converge both patterns (unanchored, both lengths, case-insensitive) and add a shared test vector set. Consider a single source of truth to prevent future drift.

### D. Robustness & coverage

- [ ] **D1 — `scrub_event_messages` fails open.** On a mid-scrub `StandardError` it returns the event unmodified, diverging from sibling `scrub_event_urls` which redacts to `[SCRUBBING_FAILED]` (fail-closed). The rescue comment even acknowledges the retained content may be sensitive. **Fix:** make it fail-closed, consistent with the URL path.
- [ ] **D2 — Transaction handler skips the route-param value layer.** `beforeSendTransaction` applies only the pattern net, not the `sentryScrubParams`-driven value layer that `beforeSend` runs first, so a param governed by route metadata (not a path pattern) leaks in the transaction name/`request.url`. Currently masked by B2. **Fix:** apply the value layer in the transaction handler too, or share one scrubbing entrypoint.
- [ ] **D3 — ~130 lines of new backend scrubbing have zero test coverage.** No spec references `scrub_transaction_event`, `scrub_event_messages`, `scrub_text`, or `scrub_query_string`; the only spec change adds `before_send_transaction` to the mock with no assertion. The parallel frontend transaction handler *did* get tests. **Fix:** add specs covering each new backend scrubber, including the fail-closed path and the identifier edge cases from C1.

## Suggested implementation approach

The cleanest shape collapses the duplicated scrubbing into a **single shared scrub entrypoint per side** (one frontend, one backend) invoked from both `beforeSend` and `beforeSendTransaction`, covering: URL, query-string param names, `request.headers` (incl. Referer), span `data` (incl. `http.query`), transaction name, and message/exception values — plus binding the frontend client via `setCurrentClient`. That structurally prevents the "one handler scrubs field X, the other forgets it" class of bug (A1–A4, D2) rather than patching each field individually.

## Definition of done

- [ ] A1–A4, B1–B2, C1, D1–D3 addressed (or explicitly deferred with rationale)
- [ ] Frontend and backend scrub the same field set; identifier patterns converge with a shared test vector set
- [ ] Frontend client is bound such that `browserApiErrorsIntegration` captures async-callback errors and transaction events are produced and scrubbed
- [ ] New backend scrubbers have unit specs, including fail-closed behavior and identifier edge cases
- [ ] Manual verification: trigger an error and a navigation on a token-bearing / identifier-bearing route with a real DSN in a test project and confirm no identifier/token appears in the captured event (url, headers, query, transaction, message)

## Testing strategy

- **Backend:** RSpec unit specs for each new scrubber (`scrub_transaction_event`, `scrub_event_messages`, `scrub_text`, `scrub_query_string`, Referer header path, identifier edge cases).
- **Frontend:** extend the existing `beforeSend`/`beforeSendTransaction` tests to cover `request.headers`, `http.query`, query-param names, and async-callback capture.
- **End-to-end:** point at a throwaway Sentry project, reproduce the flows, inspect the delivered event payloads.

## Labels

`security`, `privacy`, `frontend`, `backend`, `testing`

## Review provenance

Findings from a `max`-effort multi-agent review (24 agents: 6 finders → 15 independent verifiers → sweep → synthesis) of `git diff main...HEAD`. 19 candidates, 2 refuted (an `EMAIL_PATTERN` ReDoS claim — verified linear; and a `scrub_query_string` no-op claim — refuted). Suggested branch: `feature/<ISSUE_NUMBER>-sentry-scrub-hardening` off `main` (create via `/gh:issue-start`).

### LINKED ISSUE #3794: Sentry scrub hardening (#3788): 9 verified redaction leaks + observability regression + cleanups
## Summary

Max-effort code review of the Sentry scrub-hardening work (branch `fix/3787-sentry-scrub-hardening`, PR #3788, base `269db1d0d`) surfaced **9 verified defects**. The dominant theme is **incomplete redaction**: sensitive values (secret identifiers, emails, tokens/passphrases) still reach Sentry through several paths this PR was meant to close, plus one observability regression and three maintainability cleanups.

Six are correctness defects (five CONFIRMED, one PLAUSIBLE); three are cleanups. Three additional candidates were refuted during verification and are excluded.

Review scope: `git diff qa/sentry-settings...HEAD`, 8 changed files. Verification independently confirmed line numbers, regex behavior (empirical `node` diffs), and `@sentry/core` internals.

---

### C1 — Leading `?` defeats first sensitive query param (CONFIRMED)
**File:** `src/plugins/core/enableDiagnostics.ts:344` (parser in `src/plugins/core/diagnostics/scrubbers.ts`)

Sentry stores span `http.query` as `parsedUrl.search`, which **includes the leading `?`** (`@sentry/core` fetch.js; `@sentry/browser` tracing/request.js). `scrubSensitiveQueryParams` splits on `&` and reads the first param's name as `?<name>` (e.g. `?token`), which never matches `SENSITIVE_QUERY_PARAMS`, so the first value survives. The comment at `scrubbers.ts:340` wrongly claims `http.query` is "a bare query string."

**Leak:** With tracing on by default (`tracesSampleRate ?? 0.01`) and `setCurrentClient` (added in this PR) now letting `browserTracingIntegration` emit transactions, a request to `/api/v2/...?token=hunter2&x=1` yields `http.query = '?token=hunter2&x=1'`; `?token` is unmatched so `hunter2` ships unredacted. Affects any sensitive-named param (`token`/`key`/`secret`/`passphrase`) appearing **first** whose value is not id/email-shaped (short tokens, passphrases, custom keys) — the ID/email nets also miss those.

**Fix direction:** Strip a leading `?` before splitting query params, or parse via `URLSearchParams`.

### C2 — Backend Referer scrubbing skips the email/verifiable-ID nets (CONFIRMED)
**File:** `lib/onetime/initializers/setup_diagnostics.rb:378`

The Referer header routes through `scrub_url`, which only redacts sensitive **path** prefixes (`/secret`, `/receipt`, …) and the four named query params (`key`/`secret`/`token`/`passphrase`). Unlike the frontend Referer path (`scrubUrlWithPatterns`) and unlike backend `scrub_text`, it applies **no** `EMAIL_PATTERN` and no free-text verifiable-ID pass.

**Leak:** With `send_default_pii` enabled, an error event carrying `Referer: https://host/dashboard?ref=<62-char-secret-key>` or `.../verify?email=user@example.com` is sent to Sentry unredacted — `ref=`/`email=` are not in `SENSITIVE_QUERY_PARAMS` and no id/email net runs. `request.url` is protected only because it is a sensitive-prefixed path; the arbitrary Referer URL is not.

**Fix direction:** Run the email + verifiable-ID text nets over the Referer value (mirror `scrub_text` / the frontend `scrubUrlWithPatterns` pass).

### C3 — `\b`-anchoring drops 62-char IDs abutting word characters (CONFIRMED)
**File:** `src/plugins/core/diagnostics/scrubbers.ts:53`

`VERIFIABLE_ID_PATTERN` changed from the unanchored `/[0-9a-z]{62}/gi` to a `\b`-word-boundary-anchored pattern. A 62-char base62 secret glued to adjacent word chars no longer matches.

**Leak (empirical `node` regex diff):** `?ref=<62>abc`, `<62>x`, and `load <62>_meta` all go redacted→**LEAK** under the new pattern, while `/`-delimited and space-delimited placements stay covered. Used by both `scrubUrlWithPatterns` (pass 4) and `scrubSensitiveStrings`, so URLs, breadcrumbs, exception messages, and span descriptions are all affected.

**Note:** partially offset — the same change newly adds 31-char legacy-ID coverage the old 62-only pattern missed. The fix must keep the 31-char win while restoring unanchored 62-char matching (e.g. anchor only the left side, or drop `\b` and keep a length-based match).

### C4 — Backend transaction-span query/url skips email/ID nets (CONFIRMED)
**File:** `lib/onetime/initializers/setup_diagnostics.rb:499`

`scrub_query_string` (and `data['url']` via `scrub_url`) omits the email and verifiable-ID nets that the frontend's `scrubQueryStringValues`/`scrubUrlWithPatterns` apply.

**Leak:** With `traces_sample_rate=0.1`, a sampled span carrying `http.query='contact=user@example.com'` (or `url='https://host/download/<62-char-id>'` under a path prefix not in `IDENTIFIER_PATH_PATTERN`) ships intact — only the four param **names** are redacted. Verified: `scrub_query_string('contact=user@example.com')` and `scrub_query_string('ref='+62*'a')` both return unchanged. The frontend counterpart redacts both (`contact=[EMAIL_REDACTED]`, `ref=[REDACTED]`), so **the PR's claimed FE/BE parity is broken for span query data, with no backend test covering the email/bare-id value case.**

**Fix direction:** Apply the email + verifiable-ID nets in `scrub_query_string`/`scrub_url` value handling; add backend tests for the email and bare-id cases.

### C5 — `setCurrentClient` ships auto-captured events without deployment tags (CONFIRMED — observability regression)
**File:** `src/plugins/core/enableDiagnostics.ts:501`

`setCurrentClient(client)` now routes globally/auto-captured events (unhandled rejections, `browserApiErrors` async callbacks, `browserTracing` transactions) through the **current** scope, but `service`/`site_host`/`jurisdiction` tags are set only on the detached isolated `scope` object (lines 505/511/540), which is passed explicitly only for manual captures.

**Impact:** On multi-region / custom-domain deploys these newly-captured events reach the shared frontend Sentry project **untagged** — ops cannot filter or attribute them to deployment/region. Backend sets these via global `Sentry.set_tags`, so the asymmetry is FE-only. Before this PR the events were dropped entirely, so the untagged flow is a new regression from the binding.

**Fix direction:** Set the deployment tags on the current/global scope (`getCurrentScope().setTag(...)`) so integration-captured events carry them.

### C6 — Transaction value-layer reads global `currentRoute` (PLAUSIBLE)
**File:** `src/plugins/core/enableDiagnostics.ts:316`

`createBeforeSendTransactionHandler` calls `collectCurrentRouteValues(router)`, which reads `router.currentRoute.value` — the **current** route, not the route the transaction belongs to.

**Leak:** A pageload transaction for `https://host/page/abc123` (param marked sensitive via metadata, short enough that the always-on pattern net misses it) can still be in flight when the user navigates to `/dashboard`; `currentRoute` is now `/dashboard` with no params, so `collectCurrentRouteValues` returns `[]` and the value layer redacts nothing — `abc123`, the exact case the value layer exists to catch, ships in the transaction event. Symmetric over-redaction is the milder counterpart.

**Fix direction:** Resolve the route from the transaction's own `request.url`/name rather than the live router.

---

### Q1 — Shared-entrypoint doc block misattached (CONFIRMED)
**File:** `src/plugins/core/enableDiagnostics.ts:112`

The doc block describing the "SINGLE shared scrub entrypoint" (Covers: url + Referer + transaction) sits above `scrubRefererHeader`, which only scrubs the header. The real entrypoint `scrubCommonEventFields` (line 127) is undocumented. A maintainer trusting the comment edits/calls the wrong function and a field silently ships unscrubbed.

### Q2 — `collectCurrentRouteValues` duplicates inline route-value collection (CONFIRMED)
**File:** `src/plugins/core/enableDiagnostics.ts:80`

The new helper and the inline block at lines 176-191 in `createBeforeBreadcrumbHandler`'s `scrubNavigationUrl` both gate on `sentryScrubParams !== false`, guard `Object.keys(params).length > 0`, and call `collectValuesToRedact(params, sentryScrubParams)` — differing only in route source. A future opt-out change applied to one copy drifts breadcrumb vs event/transaction scrubbing. Extract a single helper taking a resolved-route-like `{meta, params}`.

### Q3 — New spec file instead of extending existing (PLAUSIBLE — convention)
**File:** `spec/unit/onetime/initializers/sentry_url_scrubbing_spec.rb:15`

`TESTING.md` (imported by `~/.claude/CLAUDE.md`): *"Maintain and reuse existing RSpec context files. Add new ones only when necessary for coverage."* `spec/unit/onetime/initializers/setup_diagnostics_spec.rb` already describes `Onetime::Initializers::SetupDiagnostics` with a `before_send hook` context, yet this diff opens a second `RSpec.describe` for the same class. Scrub coverage is now split across two files with duplicated setup scaffolding.

---

## Refuted during verification (excluded)

- `setup_diagnostics.rb:537` — rescue-block "duplication" in `scrub_event_messages`: real overlap, but the rescue correctly redacts both fields; no observable defect.
- `setup_diagnostics.rb:391` — `redact_url_bearing_headers` vs `scrub_url_bearing_headers`: shared frozen `URL_BEARING_HEADERS` constant; stated failure scenario factually wrong.
- `enableDiagnostics.ts:501` — globalHandlers double-capture of Vue `errorHandler` errors: Sentry dedupes.

---

## Acceptance criteria

- [ ] **C1** — a sensitive param appearing first in `http.query` (with the leading `?`) is redacted; regression test with `?token=...&x=1`.
- [ ] **C2** — Referer values carrying an email or 62/31-char verifiable ID under a non-sensitive path/param are redacted; backend test added.
- [ ] **C3** — 62-char IDs abutting word chars (`<id>abc`, `<id>x`, `load <id>_meta`) are redacted again while the new 31-char coverage is retained; regex unit tests for both.
- [ ] **C4** — backend span `http.query`/`url` values redact emails and bare verifiable IDs, restoring FE/BE parity; backend tests for email and bare-id cases.
- [ ] **C5** — integration-captured FE events (unhandled rejection, async callback, transaction) carry `service`/`site_host`/`jurisdiction` tags.
- [ ] **C6** — the transaction value-layer resolves the transaction's own route, not the live router (or the risk is explicitly documented as accepted).
- [ ] **Q1/Q2/Q3** — doc block reattached to `scrubCommonEventFields`; route-value collection unified into one helper; scrub specs consolidated into `setup_diagnostics_spec.rb`.

## Testing strategy

- Frontend: extend `scrubbers.spec.ts` with the C1/C3 query-param and ID-anchoring cases; add a transaction-tag assertion for C5.
- Backend: extend `setup_diagnostics_spec.rb` (see Q3) with C2/C4 Referer and span-query email/id cases.
- Verify the existing suites (248 FE / 120 BE per the branch note) stay green.

## Suggested branch

`feature/<ISSUE_NUMBER>-sentry-scrub-leak-hardening` from `main` — not created; use `/gh:issue-start` when ready.

### PR DESCRIPTION
**Pull request by delano**
## Summary

Hardens Sentry/diagnostics data scrubbing across the frontend and backend so sensitive values (emails, secret identifiers, query params, URL nets, headers) are redacted before events leave the client. Binds the isolated diagnostics client so async errors and transaction scrubbing actually take effect, and adds fail-closed redaction for messages/exceptions.

Backend `setup_diagnostics.rb` gains root-tier redaction with the email-before-ID net ordering and unanchored 62-char identifier matching; frontend `scrubbers.ts` / `enableDiagnostics.ts` mirror the same scrub tiers with nullish-passthrough typing. Replaces the old `sentry_url_scrubbing_spec` with a comprehensive `setup_diagnostics_spec` (849 lines) and adds FE scrubber specs plus backend/frontend diagnostics docs.

## Testing

- Backend: RSpec green
- Frontend: vitest green

Resolves #3787
Resolves #3794

### CODE REVIEW

File: src/plugins/core/enableDiagnostics.ts [hunk 1]
```diff
@@ -53,6 +68,127 @@ function scrubUrlValuesThenPatterns(url: string, sortedValues: string[]): string
...
+      // router.resolve wants an in-app location, not a full URL. Use a
+      // synthetic base so bare paths parse too; the base is discarded.
+      const parsed = new URL(url, 'http://_');
+      const resolved = router.resolve(parsed.pathname + parsed.search + parsed.hash);
+      return collectRouteParamValues(resolved as ResolvedRouteLike);
+    } catch {
+      // Fall through to the current-route fallback below.
+    }
+  }
+  return collectCurrentRouteValues(router);
```
[the remark this answers is not in the document]
Comment (delano) [reply src/plugins/core/enableDiagnostics.ts]: Good catch — confirmed reachable and fixed in 6120956e4.

Scrubbing defaults to scrub-all-params and the path net only covers 6 prefixes (`secret|private|receipt|incoming|invite|confirm`), so the value layer is the primary protection for params on other routes. On the no-URL / unresolvable-URL fallback, reading the live current route after a navigation did borrow the wrong page's params — missing the transaction's own short param and implying coverage that wasn't there.

Fix: the current-route fallback now only applies when the current route IS the transaction's route (`event.transaction === currentPath`, the parameterized `matched.at(-1).path`). On a navigate-away mismatch it contributes no values and the always-on pattern/path nets are the floor. The preferred path (resolve the transaction's own route from `request.url`) is unchanged.

Residual, documented in-code: a short, meta-scrubbed param on a non-sensitive-path route is unrecoverable at the value layer once its URL is gone *and* the user has navigated away — only the nets apply. That's inherent (no route context to map the opaque value to "sensitive"); the fix stops it from being *silently* half-covered by unrelated route state. New test `does NOT borrow current-route params after navigating to a different route` asserts the guard bites and the net floor still redacts known shapes.
